applications/OptimizationApplication/python_scripts/algorithms/algorithm_gradient_projection.py

```diff
@@ -1,120 +1,228 @@
+        self.algorithm_data.GetBufferedData()["search_direction"] = search_direction.Clone()
+        self.algorithm_data.GetBufferedData()["correction"] = correction.Clone()
+
+    def __CollectiveListCollectiveProduct(self, collective_list: 'list[KratosOA.CollectiveExpression]', other_collective: KratosOA.CollectiveExpression) -> Kratos.Vector:
+        result = Kratos.Vector(len(collective_list))
+
+        for i, collective_list_item in enumerate(collective_list):
+            result[i] = KratosOA.ExpressionUtils.InnerProduct(collective_list_item, other_collective)
+        return result
+
+    def __CollectiveListVectorProduct(self, collective_list: 'list[KratosOA.CollectiveExpression]', vector: Kratos.Vector) -> KratosOA.CollectiveExpression:
+        if len(collective_list) != vector.Size():
+            raise RuntimeError(f"Collective list size and vector size mismatch. [ Collective list size = {len(collective_list)}, vector size = {vector.Size()}]")
+        if len(collective_list) == 0:
+            raise RuntimeError(f"Collective lists cannot be empty.")
+
+        result = collective_list[0].Clone() * 0.0
+        for i, collective_list_item in enumerate(collective_list):
+            result += collective_list_item * vector[i]
+
+        return result
+    
+    @time_decorator()
+    def ComputeControlUpdate(self, alpha) -> KratosOA.CollectiveExpression:
+        update = self.algorithm_data.GetBufferedData()["search_direction"] * alpha + self.algorithm_data.GetBufferedData()["correction"]
+        self.algorithm_data.GetBufferedData()["control_field_update"] = update.Clone()
+
+    @time_decorator()
+    def UpdateControl(self) -> KratosOA.CollectiveExpression:
+        update = self.algorithm_data.GetBufferedData()["control_field_update"]
+        self.__control_field += update
+        self.algorithm_data.GetBufferedData()["control_field"] = self.__control_field.Clone()
+
+    @time_decorator()
+    def GetCurrentObjValue(self) -> float:
+        return self.__obj_val
+    
+    @time_decorator()
+    def GetCurrentControlField(self):
+        return self.__control_field
+    
+    @time_decorator()
+    def Output(self) -> KratosOA.CollectiveExpression:
+        self.CallOnAllProcesses(["output_processes"], Kratos.OutputProcess.PrintOutput)
+    
+    @time_decorator()
+    def Solve(self):
+        while not self.converged:
+            with OptimizationAlgorithmTimeLogger("Gradient Projection",self._optimization_problem.GetStep()):
+                self.__obj_val = self.__objective.CalculateStandardizedValue(self.__control_field)
+                obj_info = self.__objective.GetInfo()
+                self.algorithm_data.GetBufferedData()["std_obj_value"] = obj_info["value"]
+                self.algorithm_data.GetBufferedData()["rel_obj[%]"] = obj_info["rel_change [%]"]
+                if "abs_change [%]" in obj_info:
+                    self.algorithm_data.GetBufferedData()["abs_obj[%]"] = obj_info["abs_change [%]"]
```

Comment:
I am little bit confused in here. Why do you say `abs_obj` = `abs_change`?

Reply:
it just a naming for abs change of objective. I can rename, just suggest a new name.

Reply:
```suggestion
                    self.algorithm_data.GetBufferedData()["abs_change[%]"] = obj_info["abs_change [%]"]
```